Formal2024/Workshops/workshop5.lean

```diff
@@ -0,0 +1,402 @@
+
+def double (n : ℕ) : NatEven := sorry
+
+end
+
+section
+
+/-
+  # Quotient types
+
+  Quotient types are Lean's way of dealing with types of equivalence classes under some relation.
+  These are one of the few types that are not defined in terms of inductive types. Instead, they
+  are an axiom of Lean's type theory.
+-/
+
+-- In Lean, equivalence relations are called `Equivalence`s. This is a structure that packages a
+-- relation on a type, and a proof that it is an equivalence relation.
+#check Equivalence
+
+-- A type equipped with an equivalence relation is a `Setoid`. This is defined as a `class` which
+-- is a particular kind of structure. We will learn more about classes next week.
+#check Setoid
+
+-- For example, here is a `Setoid`:
+variable (n : ℕ)
+
+def NatModRel : Setoid ℕ where
+  r := fun a b => a % n = b % n
+  iseqv := ⟨fun x => rfl,
+    fun hxy => hxy.symm,
+    fun hxy hyz => by rw [hxy, hyz]⟩
+-- This is some convenient syntax for constructing structures. We will see more about it next week.
+-- We provide a relation `r` and a proof that it is an equivalence relation `iseqv`.
+
+-- The type of equivalence classes of a `Setoid` is constructed using `Quotient`.
+def NatMod : Type := Quotient (NatModRel n)
+
+-- To produce an element of a quotient, we use `mk` or `⟦⟧`.
+#check Quotient.mk (NatModRel n)
+#check (⟦0⟧ : NatMod n)
+
+-- To define a function out of a quotient type, we use the universal property.
+-- This can be accessed with `Quotient.lift`.
+#check Quotient.lift
+
+-- `Quotient.sound` is useful for showing that two terms of the quotient are equal.
+#check Quotient.sound
+
+-- For example, we can define a sucessor function on `ℕmod n`.
+def modSucc {n : ℕ} : NatMod n → NatMod n:=
+  Quotient.lift (λ k ↦ ⟦k + 1⟧) (by
+    intro a b h
+    apply Quot.sound
+    dsimp [NatModRel]
+    rw [Nat.add_mod a, Nat.add_mod b, h]
+  )
+
+-- Lastly, `Quotient.ind` lets us prove things about terms in a quotient type if we
+-- can prove them for any representative.
+#check Quotient.ind
+
+example : (k : NatMod n) → ∃ h, k = ⟦h⟧ := by
+  apply Quotient.ind
+  intro h
+  exact ⟨h, rfl⟩
+
+-- This is a bit like an induction proof, so we can use the `induction'` tactic.
+-- This is one of the reasons `induction'` is more powerful than `induction`.
+example (k : NatMod n) : ∃ h, k = ⟦h⟧ := by
+  induction' k using Quotient.ind with h
+  exact ⟨h, rfl⟩
+-- Here we are telling `induction'` that the principle of induction we want to use
+-- is `Quotient.ind`.
+
+/-
+  # Exercises
+-/
+
+-- 5. Check the following property of `modSucc`.
+
+-- You might want to use the following.
+#check Quotient.lift_mk
+#check Quot.sound
+#check Nat.sub_add_cancel
+
+example (hn : n ≥ 1) : modSucc (⟦n - 1⟧ : NatMod n) = ⟦0⟧ := by
+  letI := NatModRel n
+  -- This line will make `Quotient.lift_mk` work. You can ignore it for now. We will see more
+  -- about it next week when we talk about typeclasses.
+  sorry
+
```

Comment:
Not sure if there is a better way, but I tried doing this exercise and at some point I had to simplify the symbol `≈` to get norm_num to worl. In case that's something we need to explain how to do.
```lean
example (hn : n ≥ 1) : modSucc (⟦n - 1⟧ : ℕmod n) = ⟦0⟧ := by
  letI := ℕmodRel n
  apply Quotient.sound
  rw [Nat.sub_add_cancel hn]
  simp only [(· ≈ ·), ℕmodRel]
  norm_num   
```

Reply:
There's a few ways you can do it, but I imagine that is something they might try. I added a comment about this.